### PR TITLE
KAFKA-16666: Migrate GroupMetadataMessageFormatter` to tools module

### PR DESCRIPTION
**Pull request by m1a2st**
https://issues.apache.org/jira/browse/KAFKA-16666

we need to migate GroupMetadataMessageFormatter from scala code to java code,and make the message format is json pattern
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.io.IOException;
+import java.io.PrintStream;
+import java.nio.ByteBuffer;
+import java.util.Objects;
+import java.util.Optional;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public abstract class AbstractGroupMetadataFormatter implements MessageFormatter {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java]: How about `ApiMessageFormatter`

File: tools/src/main/java/org/apache/kafka/tools/consumer/GroupMetadataMessageFormatter.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.NullNode;
+import com.fasterxml.jackson.databind.node.TextNode;
+
+import java.nio.ByteBuffer;
+import java.util.Optional;
+
+public class GroupMetadataMessageFormatter extends AbstractGroupMetadataFormatter {
+
+    JsonNode transferMetadataToJsonNode(ApiMessage logKey, short keyVersion) {
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/GroupMetadataMessageFormatter.java]: please add `Override`

File: tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java (hunk 1 above)
Comment (chia7712) [reply tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java]: For another, do we need two formatter for offset topic? It seems we can merge then into single one, right?

**Review by chia7712 — COMMENTED** [review]
@m1a2st thanks for this patch

File: tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java (hunk 1 above)
Comment (m1a2st) [reply tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java]: Is there a scenario who only want to see offset message or group metadata message, thus In old formatter split different message type.
I think we should reserve these different formatter

Comment (chia7712) [reply tools/src/main/java/org/apache/kafka/tools/consumer/AbstractGroupMetadataFormatter.java]: > I think we should reserve these different formatter

you are right.

File: tools/src/main/java/org/apache/kafka/tools/consumer/ApiMessageFormatter.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                .orElseGet(() -> new TextNode(UNKNOWN));
+    }
+    
+    private JsonNode readToValueNode(ByteBuffer byteBuffer, short version) {
+        return readToValueMessage(byteBuffer)
+                .map(logValue -> transferValueMessageToJsonNode(logValue, version))
+                .orElseGet(() -> new TextNode(UNKNOWN));
+    }
+    
+    protected abstract Optional<ApiMessage> readToKeyMessage(ByteBuffer byteBuffer);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/ApiMessageFormatter.java]: It seems `readToKeyMessage`  and `transferKeyMessageToJsonNode` cab be merged into single method. For example: `protected abstract Optional<JsonNode> readToKeyJson(ByteBuffer byteBuffer, short version);`

File: tools/src/main/java/org/apache/kafka/tools/consumer/ApiMessageFormatter.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    protected abstract JsonNode transferKeyMessageToJsonNode(ApiMessage message, short version);
+    protected abstract Optional<ApiMessage> readToValueMessage(ByteBuffer byteBuffer);
```
Comment (chia7712) [inline tools/src/main/java/org/apache/kafka/tools/consumer/ApiMessageFormatter.java]: ditto

**Comment by chia7712** [comment]
all failed tests pass on my local.

**Review by chia7712 — APPROVED** [review]
LGTM

**Pull request merged**